### ISSUE TITLE
Player enters tight infinite loop when live rendition manifests stop updating

### ISSUE THREAD

**Issue by skidder** [videojs/http-streaming#975]
## Description
The VideoJS player enters a tight infinite loop requesting a single rendition manifest when the rendition manifests for a live stream stops being updated. The request-rate is only limited by the network connection speed. Rates exceeding 50 requests/sec have been observed. This has the potential to overwhelm the player device and video origin.

This can occur when the input for an HLS live stream is disconnected and the encoder keeps the live stream active during a reconnect window. During the reconnect window, the video origin might continue to serve rendition manifests that don't contain any new segments. Once the input is reconnected, the video origin can resume updating the rendition manifests and playback continues. If the reconnect window expires without the input being reconnected, then the rendition manifests can be marked as ended.

While the reconnect window is open, the player iterates over the set of rendition manifests looking for a manifest containing new segments. If all of the rendition manifests have been retrieved and no updates were found, then the player becomes stuck trying the same rendition manifest endlessly.

## Sources
HLS live streams that support input-stream reconnect-windows.

## Steps to reproduce
Demonstration: [https://youtu.be/bGCVSud2ZGk](https://youtu.be/bGCVSud2ZGk)

1. Start sending input stream to encoder configured with a reconnect window (e.g. 5 minutes)
2. Start playback in VideoJS
3. Disconnect input stream
4. Observe the player iterate over rendition manifests
5. Once all rendition manifests have been exhausted, the player enters a tight loop requesting the same rendition manifest endlessly
6. Reconnect the input stream
7. Observe the player resume playback

### Expected
The player should continually cycle through the set of rendition manifests with some delay. This is typically enforced through the `blacklistDuration`.

### Error output
`VIDEOJS: WARN: Problem encountered with the current HLS playlist. Playlist no longer updating. Switching to another playlist.`

### videojs-http-streaming version
videojs-http-streaming 2.2.2
possibly more

### videojs version
VideoJS 7.6.6
VideoJS 7.8.4
possibly more

### Browsers
* Chrome 86.0.4240.75 (64-bit) on Windows 10
* Edge 86.0.622.43 (64-bit) on Windows 10

### Platforms
* Consistently reproducible on Windows 10
* Occasionally reproducible on Mac OS X 10.15.6

### Other Plugins
No

### Other JavaScript
No

**Comment by gkatsev** [comment]
From @gesinger in slack:

thanks for reporting, and for the PR. We had  a bit of discussion that we’ve meant to post on the issue, and the infinite requesting is definitely a problem, but for a different reason than the PR. The change in the PR might cause issues in other cases, as if the manifest updates, we have to reference the new media, or else some parts of the code might reference the old pre-update media. But, you’re right that the infinite requests are a problem. This is a symptom of the fact that the HLS spec says that if there’s no EXT-X-ENDLIST in a playlist, then it isn’t done updating, and should still be considered live. So we keep requesting, even if there’s no update, as we expect one eventually. If every media stream goes “down,” but the playlist still exists, then we should keep requesting. So for a single media playlist, this will retry that playlist. But we shouldn’t be requesting so quickly. We are going to implement an incremental backoff feature for manifest update requests when we encounter a case where none of the manifests are updating. That should slow down the requests until, in theory, it’s a long enough wait that it is as good as stopped requesting.

**Comment by gabek** [comment]
Hi there!  Are there any updates for this issue?

**Comment by dylanjha** [comment]
Hi @gabek - yep. This is still happening. Would love to help get it resolved.

**Comment by heennkkee** [comment]
This isn't really a playback issue. Unless the EXT-X-ENDLIST-tag is in the playlist, the media is supposedly not finished.

In this issue there are some ideas on workarounds:
https://github.com/videojs/http-streaming/issues/924

**Comment by gabek** [comment]
Agreed that it should continue to check until `EXT-X-ENDLIST`, however, the tight loop is indeed a problem.  The incremental backoff suggestion makes sense.

**Comment by gkatsev** [comment]
yeah, we should potentially do incremental backoff, but at the very least we should be waiting the half target duration that the HLS spec calls for. I think there's a bug where in this case we just end up retrying immediately after we get a 404.

**Comment by gkatsev** [comment]
I'm looking into this. Was able to reproduce locally by getting a 2 rendition VOD playlists and removing the end list tags.
I'll probably make two fixes for this. One to get rid of the tight loop and then follow it up later with an incremental back off.

**Comment by dylanjha** [comment]
thanks for the update @gkatsev 🎉

**Comment by gkatsev** [comment]
I posted a PR with the initial fix in #1038. Please try it out and see if it's working ok for you https://deploy-preview-1038--videojs-http-streaming.netlify.app/.
The PR currently only contains the fix for delaying to half target duration each time we make the request. It doesn't have the incremental back off feature, which will come in a separate PR.

**Comment by gkatsev** [comment]
I believe this is fixed in #1038. We'll get a VHS and a Video.js release out in the next couple of days, I hope.

**Comment by dylanjha** [comment]
I told this to gary via DM, but logging here for future reference.

Back in Oct this was easy for me to repro, but a few days ago when I tried again I (strangely) couldn't get the issue to happen again.

I tried on both:

- https://philcluff.co.uk/players/videojs-7-vhs.html
- https://videojs-http-streaming.netlify.app/

My intention was to reproduce the bug again, and then try with the patch to verify that it fixes the issue. @skidder had the  same problem as me (was able to easily repro a couple months ago, but can't anymore).

On Chrome + Safariy (Mac) I was able to easily get the player into the buffereing stalled state (rendition manifest no longer getting new segments, and also no `ENDLIST` tag). I sustained that state for > 4 minutes and was no longer seeing the player spamming rendition manifest requests.

Either way, I'm really glad @gkatsev was able to repro and get a fix out. It's a bit of a mystery why neither @skidder nor I was able to repro this.

**Comment by gkatsev** [comment]
VHS 2.4.1 was just published. Video.js pre-release is coming soon.

**Comment by gkatsev** [comment]
Video.js 7.11.1 just released with this change.

**Comment by skidder** [comment]
Thank you @gkatsev for getting this fix out! I was able to confirm that the fix is working in VideoJS 7.11.1 (recording at https://youtu.be/9mlKzcQvU5U).

@dylanjha I realized that my failed attempt to reproduce the issue a few days ago was due to my demo player using a fork of VideoJS that had my proposed fix for the issue. When I retried today with VideoJS 7.6.6 the issue was consistently repeatable.

So, yeah, I think we're good here. Thanks again!
